src/network/core/os_abstraction.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @file os_abstraction.cpp OS specific implementations of functions of
+ *                          the OS abstraction layer for network stuff.
+ *
+ * The general idea is to have simple abstracting functions for things that
+ * require different implementations for different environments.
+ * In here the functions, and their documentation, are defined only once
+ * and the implementation contains the #ifdefs to change the implementation.
+ * Since Windows is usually different that is usually the first case, after
+ * that the behaviour is usually Unix/BSD-like with occasional variation.
+ */
+
+#include "stdafx.h"
+#include "os_abstraction.h"
+#include "../../string_func.h"
+
+#include "../../safeguards.h"
+
+/**
+ * Construct the network error with the given error code.
+ * @param error The error code.
+ */
+NetworkError::NetworkError(int error) : error(error)
+{
+}
+
+/**
+ * Construct the network error with the last available error.
+ */
+NetworkError::NetworkError()
+{
+#if defined(_WIN32)
+	this->error = WSAGetLastError();
+#elif defined(__OS2__)
+	this->error = sock_errno();
+#else
+	this->error = errno;
+#endif
+}
```

Comment:
This is a strange constructor. I'd more expect it to be done as part of `GetLast` instead

Reply:
I was first thinking about using the no-argument constructor to mean: get the last error. So it started out as such, but even though it's less code it's also way less clear. So I changed it, but I can definitely move the behaviour to GetLast and remove the no-argument constructor. Good point.